### PR TITLE
Ajinkya/asan

### PR DESCRIPTION
**Pull request by veriAjinkya**
This is supposed to be merged after #52. 
Ignore changes in lit tests

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,6 +8,8 @@ project(Yul2LLVM
 
 include(GNUInstallDirs)
 
+set(CMAKE_CXX_FLAGS_DEBUG -fsanitize=address)
+
```
Comment (Technius) [inline CMakeLists.txt:12]: This should check `CMAKE_BUILD_TYPE`. Furthermore, asan is both a compile time flag and a library to link against, so we need:
```cmake
add_compile_options(-fsanitize=address)
add_link_options(-fsanitize=address)
```

**Comment by Technius** [comment]
Didn't look at the whole thing yet since it seems to be mixed with the other PR.

File: CMakeLists.txt (hunk 1 above)
Comment (veriAjinkya) [reply CMakeLists.txt]: done

**Review by Technius — CHANGES_REQUESTED** [review]

File: flake.nix [hunk 2]
```diff
@@ -124,6 +124,24 @@
...
+          name="yul2llvm_dbg";
+          propagatedBuildInputs = [ final.solc_0_8_15 final.yul2llvm_cpp_dbg ];
+          checkInputs = (builtins.filter (x: x != final.yul2llvm_cpp) prevAttr.checkInputs) ++ [final.yul2llvm_cpp_dbg] ;
+          checkPhase = ''
+            echo "Executing pytest"
+            PATH="$out"/bin:"$PATH" pytest
+
+            echo
+            echo "Executing lit tests"
+            ASAN_OPTIONS=detect_leaks=0 PATH="$out"/bin:"$PATH" lit ./tests -v
```
[the remark this answers is not in the document]
Comment (veriAjinkya) [reply flake.nix:142]: Its a nix string literal, how prepend last line of it?

[the remark this answers is not in the document]
Comment (Technius) [reply flake.nix:142]: You can use `+` for string concat, e.g.:
```nix
checkPhase = ''
  export ASAN_OPTIONS=detect_leaks=0
'' + ''
  echo hi
'';
```

[the remark this answers is not in the document]
Comment (veriAjinkya) [reply flake.nix:142]: done

File: CMakeLists.txt [hunk 3]
```diff
@@ -8,6 +8,12 @@ project(Yul2LLVM
 
 include(GNUInstallDirs)
 
+if(${CMAKE_BUILD_TYPE} STREQUAL Debug)
+message(STATUS ${CMAKE_BUILD_TYPE})
+  add_compile_options(-fsanitize=address)
+  add_link_options(-fsanitize=address)
```
[the remark this answers is not in the document]
Comment (veriAjinkya) [reply CMakeLists.txt:14]: It doesn't work on mac.

File: flake.nix [hunk 4]
```diff
@@ -124,6 +124,24 @@
...
           '';
         });
+        yul2llvm_cpp_dbg = final.yul2llvm_cpp.overrideAttrs (finalAttr: prevAttr: {
+          name = "yul2llvm_cpp_dbg";
+          cmakeBuildType = "Debug";
+        });
+
+        yul2llvm_dbg = final.yul2llvm.overridePythonAttrs (prevAttr: {
+          name="yul2llvm_dbg";
+          propagatedBuildInputs = [ final.solc_0_8_15 final.yul2llvm_cpp_dbg ];
```
[the remark this answers is not in the document]
Comment (veriAjinkya) [reply flake.nix:134]: This line adds yul2llvm_cpp_dbg instead of yul2llvm_cpp

[the remark this answers is not in the document]
Comment (veriAjinkya) [reply flake.nix:134]: Pushed that explicitly makes this replacement

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Technius) [reply CMakeLists.txt:13]: Should use `option` to explicitly declare the option: https://cmake.org/cmake/help/latest/command/option.html

**Review by Technius — APPROVED** [review]

**Pull request merged**